feat(qwik): customize error overlay

zanettin opened the pull request:
# What is it?

- [x] Feature / enhancement
- [ ] Bug
- [ ] Docs / tests

# Description

Qwikify the vite dev server error overlay. Fixes #2710.

## Current Version
![](https://user-images.githubusercontent.com/3241476/214968617-4ccf7be5-aa7b-409c-bef5-cce558e9f6b8.png)

## PR state after review session
![B1](https://user-images.githubusercontent.com/3241476/216787776-58cfc64e-63cd-4a49-88c0-c138263fa096.png)

## Terminal output (stays unaffected)
![Bildschirmfoto 2023-02-04 um 20 44 05](https://user-images.githubusercontent.com/3241476/216786757-b2c99e96-1923-474c-8420-dedc636eac5f.png)


# Info
- Used Icons are from https://www.svgrepo.com

# Todos
- [x] Get feedback regarding design
- [x] Wait till this PR https://github.com/vitejs/vite/pull/11830 gets merged
- [x] Wait till qwik uses version `4.2.0` of vite which includes the changes from above

# Follow-ups
- Remove vite code after https://github.com/vitejs/vite/pull/11970 got merged

# Checklist:

- [x] My code follows the [developer guidelines of this project](https://github.com/BuilderIO/qwik/blob/main/CONTRIBUTING.md)
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] Added new tests to cover the fix / functionality

zanettin commented on packages/qwik-city/buildtime/vite/dev-server.ts:
```diff
@@ -194,6 +194,9 @@ export function ssrDevMiddleware(ctx: BuildContext, server: ViteDevServer) {
 
       next();
     } catch (e) {
+      if (res.writableEnded) {
```
we've extended the `qwik` dev server middleware to handle errors as well instead of forwarding the to the vite error middleware. this because we are otherwise unable to ship the needed CSS to modify the overlay on SSR errors. The new error handler terminates the response with an `.end()` call. After that point it does no longer make sense to follow the `next(e)` path at all because an `header already sent` error would occur.

zanettin commented on packages/qwik/src/optimizer/src/plugins/vite-error.ts:
```diff
@@ -0,0 +1,210 @@
+import { yellow, magenta, cyan } from 'kleur/colors';
+import strip from 'strip-ansi';
+
+/**
```
i think about creating another PR on the vite repo to have an option to pass in some custom markdown into the vite error handler so we can remove the whole logic again on our side.

zanettin commented on packages/qwik/src/optimizer/src/plugins/vite-error.ts:
```diff
@@ -0,0 +1,210 @@
+}
+
+vite-error-overlay::part(tip):before {
+  content: "Not sure how to solve this? Visit https://qwik.builder.io or connect with the community on Discord.";
```
just an idea. please let me know what u think about it...

manucorporat replied:
yeah if this works! i think it's ok!

zanettin replied:
it does but w/o a link 😄 and can't style the url

manucorporat commented:
THIS IS AMAZING!! love it, this will deserve a good tweet when we merge it

shairez commented:
great job @zanettin !

Looks much more readable

BTW, for errors, I'd love to see the color "red" because it makes it easier to scan for it in the console.

Plus, not sure if it's possible, but syntax highlighting would also be great, but that's not a much

manucorporat commented:
I dont think code highlighting is an easy thing to add, i would not block this PR for that, we never had it before, but i would say, i would like to see the code with a more vibrant color, the yellow for vite was cooler.

manucorporat commented:
My only take would be to add a little bit of color, the red @shairez suggested would be nice and some other

zanettin commented:
thanks a lot @manucorporat and @shairez for your feedbacks 🙏 

the concern @shairez mentioned was about the terminal output which would stay unaffected by now.
i'll try to add some more colors to the overlay and post an update asap 🙏 

regarding the formatting/highlighting: yes we are limited as long as we'd like to keep the vite error overlay web component since the error is rendered within a `<pre>` tag. if we would like to do everything on our side (other PR imo as well) we could do that which would give us complete flexibility 👍

zanettin commented:
@manucorporat @shairez here are some other variations 👨‍🎨 😄  
pls let me know in which direction i should go or if one of them match your expectations 🙏  i've also updated the description of the PR and added a screenshot of the terminal output.

# B1
![B1](https://user-images.githubusercontent.com/3241476/216787776-58cfc64e-63cd-4a49-88c0-c138263fa096.png)

# B2
![B2](https://user-images.githubusercontent.com/3241476/216787780-29728a26-4596-4fc7-8ce6-b1efd7a69f85.png)

# B3
![B3](https://user-images.githubusercontent.com/3241476/216787790-afbfdd79-819c-468f-ade8-01e4b3ffbb5b.png)

# C1
![Bildschirmfoto 2023-02-04 um 21 26 55](https://user-images.githubusercontent.com/3241476/216788342-aff5a786-3d1c-42bc-8535-429c771da373.png)

shairez commented:
thanks @zanettin !

I prefer B2

zanettin commented:
> thanks @zanettin !
> 
> I prefer B2

thanks @shairez 🙏 maybe i should play around a bit with the aggressivity of the red color? give it a bit of opacity or so 👍 
imo a bit less "qwik"-style than the other options but it is an error overlay after all 😄

shairez commented:
I think it's OK not to have the error message with the branded blue color, because it's an error after all :)

Maybe the icons could be with that blue or something, but it's not a biggie IMO

zanettin commented:
totally with u @shairez 🙏 let me try some variations tomorrow evening 👍

zanettin commented:
# B4
![Bildschirmfoto 2023-02-05 um 23 12 13](https://user-images.githubusercontent.com/3241476/216848900-2ed39e0c-2ddb-4d57-87cf-0c158e9aa28a.png)
tiny adjusted B2 version (reduced red, removed blue completely from the screen)

nnelgxorz commented:
I think it looks great. 

I personally prefer not having the error message header as red and reserving the red color for highlighting the actual error in the code, [like this](https://elm-lang.org/assets/blog/error-messages/0.16/context.png). Giant red text just makes me feel like I'm being yelled at. 

I would bet that it stresses out beginners as well. Just my two cents. 👍

shairez commented:
but you ARE being yelled at @nnelgxorz !! 😅😂

adamdbradley commented:
Love it, and like the idea of less red and more qwik-ish colors. But think we ship it and see how it feels and iterate.

zanettin commented:
thanks for all the feedbacks 🙏 set the current state of this PR to the B1 version. hope that is a good starting point. if you have other wishes/requirements, pls let me know 🙏 
otherwise i'll wait till the PR on the vite repo got merged (was approved twice now 🎉 ) and we've upgraded to latest vite version for final testings.

manucorporat commented:
B1 is cool!
